### ISSUE TITLE
Provider doesn't seem to use proxy settings

### ISSUE THREAD

**Issue by raphaelperrin24** [integrations/terraform-provider-github#1509]
Using TFE on premise behind a company proxy.

On Applying plan, we get the error below:
`Error: Post "https://api.github.com/user/repos": dial tcp 140.82.121.5:443: connect: connection timed out`

Wondering if the provider respects the setting. In the worker container, the proxy settings are set as environment variables.

**Comment by mkilchhofer** [comment]
Maybe somehot related:
- https://github.com/google/go-github/issues/2363

**Comment by kfcampbell** [comment]
The GitHub provider does not have any proxy-related code or behavior; I imagine that the answer may lie in either the go-github package or else Terraform proxy behavior as a whole. I'm going to close this issue now as there hasn't been a response in quite some time, though please feel free to reopen if you feel it's warranted.

**Comment by mkilchhofer** [comment]
@kfcampbell can you reopen this please?
The issue referenced above shows exactly how to implement this.
It has to be setup on go's http client this is reused in the library `google/go-github`.

**Comment by kfcampbell** [comment]
>  The issue referenced above shows exactly how to implement this.

There's not a linked issue above. What issue are you referring to? 

>  It has to be setup on go's http client this is reused in the library google/go-github.

What does that mean about the provider's involvement? It seems unrelated.

**Comment by smittnacht** [comment]
Hi @kfcampbell,
I agree with the assessment from @mkilchhofer and @raphaelperrin-swisspost that the current implementation doesn't adhere to proxy settings when making requests to GitHub.

Upon examining the go-github project, which is responsible for creating the actual HTTP client, we find two methods for client creation: NewClient() and NewClientWithEnvProxy() (introduced in https://github.com/google/go-github/pull/2686). Since you're employing the NewClient() method without explicitly supplying any proxy-related configuration, no proxy is being utilized at all.

I believe a straightforward solution would be to simply replace NewClient() with NewClientWithEnvProxy(). This would ensure that proxy settings are respected when making requests to GitHub.

Please share your thoughts on this approach.

**Comment by mkilchhofer** [comment]
> There's not a linked issue above. What issue are you referring to? 

This one here:
- https://github.com/google/go-github/issues/2363

> What does that mean about the provider's involvement? It seems unrelated. 

See detailed answer of @smittnacht

**Comment by smittnacht** [comment]
@mkilchhofer @kfcampbell could someone of you or any other maintainer reopen this issue please?

**Comment by kfcampbell** [comment]
Reopened; sorry about that. I'm receptive to ideas or PRs are about how to fix this!

**Comment by mkilchhofer** [comment]
Not stale

**Comment by ytausch** [comment]
Related to this issue:

If anyone wants to use Burp as a proxy for easier debugging, I've used the invisible proxy feature of Burp to circumvent the current limitations of missing proxy support: [gist](https://gist.github.com/ytausch/2add419ff03ccf7a9eba45cd39fe7714)

**Comment by ytausch** [comment]
not stale

**Comment by bufferoverflow** [comment]
My colleague @fh1ch made it work with this patch:

```diff
diff --git a/github/config.go b/github/config.go
index de232c17..f71fc577 100644
--- a/github/config.go
+++ b/github/config.go
@@ -73,7 +73,8 @@ func RateLimitedHTTPClient(client *http.Client, writeDelay, readDelay, retryDela
 }

 func (c *Config) AuthenticatedHTTPClient() *http.Client {
-       ctx := context.Background()
+       customHttpClient := &http.Client{Transport: &http.Transport{Proxy: http.ProxyFromEnvironment}}
+       ctx := context.WithValue(context.Background(), oauth2.HTTPClient, customHttpClient)
        ts := oauth2.StaticTokenSource(
                &oauth2.Token{AccessToken: c.Token},
        )
@@ -87,7 +88,7 @@ func (c *Config) Anonymous() bool {
 }

 func (c *Config) AnonymousHTTPClient() *http.Client {
-       client := &http.Client{Transport: &http.Transport{}}
+       client := &http.Client{Transport: &http.Transport{Proxy: http.ProxyFromEnvironment}}
        return RateLimitedHTTPClient(client, c.WriteDelay, c.ReadDelay, c.RetryDelay, c.ParallelRequests, c.RetryableErrors, c.MaxRetries)
 }

```

**Comment by deiga** [comment]
@bufferoverflow @fh1ch Would you be interested in opening a PR with those changes and some tests & documentation?

**Comment by bufferoverflow** [comment]
@deiga Yes, we will do. But @fh1ch is on vacation in I'm on they way to, so it will take a little while, sorry!

**Comment by dvcorreia** [comment]
I looked into this and for authenticated requests, the proxy seems to be respected. In the `AuthenticatedHTTPClient`, `oauth2.NewClient(ctx, ts)` call, if no http client is passed in the context, the [`http.DefaultClient` will be used](https://cs.opensource.google/go/x/oauth2/+/master:internal/transport.go;l=27), which respects the proxy settings.

Where I found that this is not the case is for unauthenticated requests using the `AnonymousHTTPClient`. The change for `AnonymousHTTPClient ` proposed in https://github.com/integrations/terraform-provider-github/issues/1509#issuecomment-3875958304 works, but we can also reuse `http.DefaultClient` or:

```diff
diff --git a/github/config.go b/github/config.go
index 0c5f1827..c17a7e2e 100644
--- a/github/config.go
+++ b/github/config.go
@@ -87,7 +87,7 @@ func (c *Config) Anonymous() bool {
 }
 
 func (c *Config) AnonymousHTTPClient() *http.Client {
-	client := &http.Client{Transport: &http.Transport{}}
+	client := &http.Client{Transport: http.DefaultTransport.(*http.Transport).Clone()}
 	return RateLimitedHTTPClient(client, c.WriteDelay, c.ReadDelay, c.RetryDelay, c.ParallelRequests, c.RetryableErrors, c.MaxRetries)
 }
```

**Comment by deiga** [comment]
Resolved by #3326
